lib/src/nullable_iterator.dart

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
```

Comment:
We should instead migrate our uses of the iterator to not rely on a `null` current value. It should instead use the return value of the last `moveNext` call to know when the iterator is done.

Reply:
Here's the thing, the code sometimes advances the current Iterator pointer after calling `_locateResource`:
```dart
    var entity = await _locateResource('.', iterator..moveNext());
```
With the line above, if the iterator only had one item, `iterator.current` would still return `null` before 
`moveNext` is called after the method executes.

Also, `moveNext` moves the pointer every time it gets called, which means that there's no method to peek the `Iterator` as to whether it has another item. Would it make sense to create an iterator that remembers the last time `moveNext` was called, or if it has items to iterate (i.e. a `hasNext` method)?

Please advise on how you want to proceed with this.

Reply:
Ya, dealing with iterators that were implemented via `null` checks on `current` is a pain for sure.

You may end up wanting to restructure the code some to make it easier, but generally what I have done in the past is:

- Create some variable in the same scope as the iterator, `var hasNext = true;`
- All calls to `moveNext` should be guarded with an `if (hasNext)` check, and should assign their result to `hasNext`.
- All null checks on `current` should also change to `hasNext` checks.

This is sometimes easier said than done and may need some restructuring though.

Reply:
Added a new type of `Iterator` as per your suggestion. Hopefully it should suffice.